src/hypothesis/internal/conjecture/engine.py

```diff
@@ -253,6 +276,74 @@ def test_function(self, data):
         if self.__tree_is_exhausted():
             self.exit_with(ExitReason.finished)
 
+        self.record_for_health_check(data)
+
+    def record_for_health_check(self, data):
+        # Once we've actually found a bug, there's no point in trying to run
+        # health checks - they'll just mask the actually important information.
+        if data.status == Status.INTERESTING:
+            self.health_check_state = None
+
+        state = self.health_check_state
+
+        if state is None:
+            return
+
+        state.draw_times.extend(data.draw_times)
+
+        if data.status == Status.VALID:
+            state.valid_examples += 1
+        elif data.status == Status.INVALID:
+            state.invalid_examples += 1
+        else:
+            assert data.status == Status.OVERRUN
+            state.overrun_examples += 1
+
+        max_valid_draws = 10
+        max_invalid_draws = 50
+        max_overrun_draws = 20
+
+        assert state.valid_examples <= max_valid_draws
+
+        if state.valid_examples == max_valid_draws:
+            self.health_check_state = None
+            return
+
+        if state.overrun_examples == max_overrun_draws:
+            fail_health_check(self.settings, (
+                'Examples routinely exceeded the max allowable size. '
+                '(%d examples overran while generating %d valid ones)'
+                '. Generating examples this large will usually lead to'
+                ' bad results. You should try setting average_size or '
```

Comment:
IIRC we no longer want to recommend using `average_size`; good time to take it out of the message?